src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -1556,106 +1523,284 @@ internal static string CheckSuspiciousContent(Ast scriptBlockAst)
-            "DefinePInvokeMethod", "GetTypes", "GetAssemblies", "Methods", "Properties",
+                return null;
+            }
 
-            // Suspicious methods / properties on "Type"
-            "GetConstructor", "GetConstructors", "GetDefaultMembers", "GetEvent", "GetEvents", "GetField",
-            "GetFields", "GetInterface", "GetInterfaceMap", "GetInterfaces", "GetMember", "GetMembers",
-            "GetMethod", "GetMethods", "GetNestedType", "GetNestedTypes", "GetProperties", "GetProperty",
-            "InvokeMember", "MakeArrayType", "MakeByRefType", "MakeGenericType", "MakePointerType",
-            "DeclaringMethod", "DeclaringType", "ReflectedType", "TypeHandle", "TypeInitializer",
-            "UnderlyingSystemType",
+            /// <summary>
+            /// Scan a string for suspicious content.
+            ///
+            /// This is based on the Rubin-Karp algorithm, but heavily
+            /// modified to support searching for multiple patterns at
+            /// the same time.
+            ///
+            /// The key difference from Rubin-Karp is that we don't undo
+            /// the hash of the first character as we shift along in the
+            /// input.
+            ///
+            /// Instead, we can rely on knowing we need the hashes for
+            /// shorter strings anyway, so we reuse their values in
+            /// computing the hash for the longer patterns. This lets us
+            /// use a much simpler hash as well - we can avoid the use of
+            /// mod.
+            /// </summary>
+            /// <returns>The string matching the hash, or null.</returns>
+            public static string Match(string text)
+            {
+                // The longest pattern is 29 characters.
+                // The values in the array are the computed hashes of length
+                // index-1 (so runningHash[0] holds the hash for length 1).
+                var runningHash = new uint[29];
+
+                int longestPossiblePattern = 0;
+                for (int i = 0; i < text.Length; i++)
+                {
+                    uint h = text[i];
+                    if (h >= 'A' && h <= 'Z')
+                    {
+                        h = h | 0x20; // ToLower
+                    }
+                    else if (!((h >= 'a' && h <= 'z') || h == '-'))
+                    {
+                        // If the character isn't in any of our patterns,
+                        // don't bother hashing and reset the running length.
+                        longestPossiblePattern = 0;
+                        continue;
+                    }
 
-            // Doing things with System.Runtime.InteropServices
-            "InteropServices", "Marshal", "AllocHGlobal", "PtrToStructure", "StructureToPtr",
-            "FreeHGlobal", "IntPtr",
+                    for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
```

Comment:
I must be missing something because I don't understand this loop.  It looks like computing the next hash (based on HashNewPattern below) simply takes the previous hash, multiples by LCG, and adds next character.  Why are previous hashes re-computed (j-1, j=(i-1) -> 1)?

Reply:
I'm not sure how to improve the comment. I'll try again I guess.

Briefly - omitting many other steps:

Iteration n: compute hash on `Emi` (len 3)
Iteration n+1: we compute hash on `Emit` (len 4) using hash from previous iteration (`j-1`)
Iteration n+1: compute hash on `mit` (len 3) (overwriting previous iteration, so that's why we go from longest to shortest.)

Reply:
Ok, I understand the progression now and the loop makes sense.  This makes the algorithm much more clear (at least to me).  Can you add this to your comment?